[MPS] Fix `torch.[all|any]` for 5+D tensors

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130542
* #130541

Workaround bug in `reductionAndWithTensor:` that kills app with the
following assert if 5+D tensor as an input
```
Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function encodeNDArrayOp, file GPUReductionOps.mm, line 76.
```
by reshaping the tensor to 2D/3D one before running the reduction.

Refactored common code into `all_any_common_impl_mps` as both `reductionOrWithTensor:` and `reductionAndWithTensor:` suffer from the same issue

Enabled `test_reduction_ops_5D` and  added regression test to it

malfet committed:
[MPS] Fix `torch.all` for 5+D tensors

Workaround bug in `reductionAndWithTensor:` that kills app with the
following assert if 5+D tensor as an input
```
Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function encodeNDArrayOp, file GPUReductionOps.mm, line 76.
```
by reshaping the tensor to 2D/3D one before running the reduction

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Fix `torch.all` for 5+D tensors"


Workaround bug in `reductionAndWithTensor:` that kills app with the
following assert if 5+D tensor as an input
```
Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function encodeNDArrayOp, file GPUReductionOps.mm, line 76.
```
by reshaping the tensor to 2D/3D one before running the reduction.

Add regression test to https://github.com/pytorch/pytorch/issues/95538

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/mps/operations/ReduceOps.mm:
```diff
@@ -1206,14 +1206,37 @@ Tensor std_mps(const Tensor& input_t,
+      if (input_t.ndimension() > 4) {
+        auto reduceDimLen = input_t.size(dim_);
+        if (dim_ == 0) {
+          castInputTensor = [mpsGraph reshapeTensor:castInputTensor withShape:@[ @(reduceDimLen), @-1 ] name:nil];
```
I guess you can always do this reshape because the Tensor here is always contiguous?

malfet replied:
To be frank, I don't know how MPS does reshapes under the hood, but currently there is no way to define a non-contiguous MPS tensor.

malfet committed:
Update on "[MPS] Fix `torch.[all|any]` for 5+D tensors"


Workaround bug in `reductionAndWithTensor:` that kills app with the
following assert if 5+D tensor as an input
```
Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function encodeNDArrayOp, file GPUReductionOps.mm, line 76.
```
by reshaping the tensor to 2D/3D one before running the reduction.

Add regression test to https://github.com/pytorch/pytorch/issues/95538

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Fix `torch.[all|any]` for 5+D tensors"


Workaround bug in `reductionAndWithTensor:` that kills app with the
following assert if 5+D tensor as an input
```
Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function encodeNDArrayOp, file GPUReductionOps.mm, line 76.
```
by reshaping the tensor to 2D/3D one before running the reduction.

Refactored common code into `all_any_common_impl_mps` as both `reductionOrWithTensor:` and `reductionAndWithTensor:` suffer from the same issue

Enabled `test_reduction_ops_5D` and  added regression test to it

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good to me @kulinseth can you take a final look?

malfet commented:
@pytorchbot merge -f "Lint and MPS are green"
